### PR TITLE
Use ctest instead of `make validate`

### PR DESCRIPTION
**Pull request by mpreisler**
#### Description:

- Instead of using a `validate` build target for validation and other tests, this PR switches over to `ctest`
- Adds a test that uses `linkchecker` to check URLs in the HTML guides
- Adds a test that uses `linkchecker` to check URLs in the HTML tables

How to test:
```
cd build/
cmake ../
make -j 4
ctest -j 4
```

#### Rationale:

- `ctest` output is more suitable for regression testing
- it more clearly separates testing and building

### CODE REVIEW

**Comment by mpreisler** [comment]
(jenkins is expected to fail because I reomved the validate build target)

Output example:
```
$ ctest -j 4
Test project /home/mpreisle/d/scap-security-guide/build
        Start  97: validate-ssg-rhel7-oval.xml
        Start  74: validate-ssg-rhel6-oval.xml
        Start  19: validate-ssg-fedora-oval.xml
        Start 130: validate-ssg-sle12-oval.xml
  1/160 Test #130: validate-ssg-sle12-oval.xml ....................................   Passed    0.42 sec
  2/160 Test  #19: validate-ssg-fedora-oval.xml ...................................   Passed    0.42 sec
        Start  12: validate-ssg-debian8-oval.xml
        Start  67: validate-ssg-rhel-osp7-oval.xml
  3/160 Test  #74: validate-ssg-rhel6-oval.xml ....................................   Passed    0.72 sec
  4/160 Test  #97: validate-ssg-rhel7-oval.xml ....................................   Passed    0.72 sec
  5/160 Test  #12: validate-ssg-debian8-oval.xml ..................................   Passed    0.12 sec
        Start 144: validate-ssg-ubuntu1604-oval.xml
        Start 137: validate-ssg-ubuntu1404-oval.xml
        Start 151: validate-ssg-wrlinux-oval.xml
  6/160 Test  #67: validate-ssg-rhel-osp7-oval.xml ................................   Passed    0.12 sec
  7/160 Test #144: validate-ssg-ubuntu1604-oval.xml ...............................   Passed    0.15 sec
  8/160 Test #151: validate-ssg-wrlinux-oval.xml ..................................   Passed    0.15 sec
        Start 123: validate-ssg-sle11-oval.xml
        Start  60: validate-ssg-opensuse-oval.xml
        Start  92: linkchecker-ssg-sl6-guides
  9/160 Test #137: validate-ssg-ubuntu1404-oval.xml ...............................   Passed    0.16 sec
 10/160 Test  #60: validate-ssg-opensuse-oval.xml .................................   Passed    0.20 sec
        Start 118: linkchecker-ssg-sl7-guides
        Start  99: linkchecker-ssg-rhel7-guides
 11/160 Test #123: validate-ssg-sle11-oval.xml ....................................   Passed    0.30 sec
        Start  89: linkchecker-ssg-centos6-guides
 12/160 Test  #92: linkchecker-ssg-sl6-guides .....................................   Passed    4.31 sec
        Start 115: linkchecker-ssg-centos7-guides
 13/160 Test #118: linkchecker-ssg-sl7-guides .....................................   Passed    4.51 sec
        Start 117: validate-ssg-sl7-ds.xml
 14/160 Test  #99: linkchecker-ssg-rhel7-guides ...................................   Passed    4.51 sec
        Start  98: validate-ssg-rhel7-ds.xml
 15/160 Test  #89: linkchecker-ssg-centos6-guides .................................   Passed    4.51 sec
        Start  76: linkchecker-ssg-rhel6-guides
...
```

**Comment by dahaic** [comment]
It will be more efficient to check all guides across all products at once, as there is big overlap of links between them.

**Comment by mpreisler** [comment]
> It will be more efficient to check all guides across all products at once, as there is big overlap of links between them.

that's a good point, will do

**Comment by mpreisler** [comment]
> It will be more efficient to check all guides across all products at once, as there is big overlap of links between them.

Done for tables and guides. I ended up with 2 tests, one tests all HTML guides and one tests all HTML tables.

```
$ ctest -R linkchecker
Test project /home/mpreisle/d/scap-security-guide/build
    Start 1: linkchecker-ssg-guides
1/2 Test #1: linkchecker-ssg-guides ...........   Passed   16.63 sec
    Start 2: linkchecker-ssg-tables
2/2 Test #2: linkchecker-ssg-tables ...........   Passed    1.41 sec

100% tests passed, 0 tests failed out of 2

Total Test time (real) =  18.04 sec
```

**Comment by matejak** [comment]
It works for me nicely. I see that code that handled dependencies (e.g. ensure datastream is built before testing it) has disappeared, but nothing reasonable can be done to bring it back: https://stackoverflow.com/questions/733475/cmake-ctest-make-test-doesnt-build-tests

**Comment by dahaic** [comment]
@mpreisler this PR is called experiment - it is to be merged and applied in CI?

**Comment by mpreisler** [comment]
> @mpreisler this PR is called experiment - it is to be merged and applied in CI?

The only thing missing here is docs update.

**Comment by mpreisler** [comment]
This is now ready for review. It will fail jenkins, once it's merged we need to update how jenkins tests.

File: cmake/SSGCommon.cmake [hunk 1]
```diff
@@ -1084,12 +984,29 @@ macro(ssg_build_html_stig_tables PRODUCT STIG_PROFILE DISA_STIG_VERSION)
...
         DESTINATION "${SSG_TABLE_INSTALL_DIR}")
     install(FILES "${CMAKE_BINARY_DIR}/tables/table-${PRODUCT}-stig-testinfo.html"
         DESTINATION "${SSG_TABLE_INSTALL_DIR}")
 endmacro()
 
+macro(ssg_define_linkchecker_tests)
+    if (SSG_LINKCHECKER_VALIDATION_ENABLED AND LINKCHECKER_EXECUTABLE)
+        add_test(
+            NAME "linkchecker-ssg-guides"
+            COMMAND "${LINKCHECKER_EXECUTABLE}" ${SSG_LINKCHECKER_GUIDE_FILE_LIST}
```
Comment (dahaic) [inline cmake/SSGCommon.cmake]: You have to use `--check-extern` for it to do what we want.

Comment (mpreisler) [reply cmake/SSGCommon.cmake]: lol

Comment (mpreisler) [reply cmake/SSGCommon.cmake]: DONE

**Comment by mpreisler** [comment]
I have fixed checking of external links and fixed the 2 URLs that were `404-ing`.

**Comment by dahaic** [comment]
ACK

**Pull request merged**